---
title: "_qcCHIP_ User's Guide"
author: |
  | Xiang Liu <xiang.liu@moffitt.org>
  | Mingxiang Teng <mingxiang.teng@moffitt.org>
  | Department of Biostatistics and Bioinformatics
  | Moffit Cancer Center, Tampa, FL, USA
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true
  html_document:
    toc: true
    df_print: paged
graphics: true
vignette: "%\\VignetteIndexEntry{qcCHIP user's guide} %\\VignettePackage{qcCHIP} %\\VignetteEncoding{UTF-8}
  %\\VignetteEngine{knitr::rmarkdown}\n"
---

```{r, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  error = FALSE,
  warning = FALSE, 
  message = FALSE,
  cache = FALSE,
  comment = "#>",
  dev="png"
)
```

# Introduction
Clonal hematopoiesis (CH) is a molecular biomarker associated with various 
adverse outcomes in healthy and disease individuals. Detecting CHs usually 
involves genomic sequencing of individual blood samples followed by rigorous 
bioinformatics data filtering. We report an R package, qcCHIP, a bioinformatics 
pipeline to identify CH from sequencing data by implementing a series of quality
control filters and permutation-based parameter optimization.

qcCHIP takes inputs in VCF format and filter CH mutations using technical, 
functional, individual and populations metrics associated to the mutations. 
It allows customized parameter settings of those metrics as well as parameter 
optimization using permutation analysis with cohort-specific characteristics. 
For more detail, please refer to our manuscript and the help pages of R functions. 

# Install package

Install _qcCHIP_ package via _devtools_.
```{r install}
library(devtools)
devtools::install_github("https://github.com/tenglab/qcCHIP.git",force=T)
```

# Getting Started

Load the package in R.

```{r library}
library(qcCHIP)
library(GenomicRanges)
```


# Preparing Input Files

qcCHIP works on mutation calls to optimize parameters and filter CHs. For 
each blood sample, mutations should be first called with blood sequencing data 
using tools like MuTect2. Usually, a VCF file will be generated by these tools 
to document mutations for a given sample. As qcCHIP is built to work on a clinical 
cohort instead of a single sample, the main functions of qcCHIP take input a merged 
mutation file from all samples for a given cohort. qcCHIP package provides an internal
function _vcf2input_ to merge multiple VCF files into an annotated text file, which will 
serve as the input of main functions. The merged mutation text file contains 
the following columns. The users can also manually built their merged mutation file.

  1. Chr: chromosome of variant. Exp: chr1, chr2,chrX.
  1. Start: start posation of variant.
  1. End: end posation of variant.
  1. Ref: reference allele.
  1. Alt: alternative allele.
  1. TLOD: TLOD or Qual Info from vcf file.
  1. SOR: SOR Info from vcf file.
  1. AD_alt: Allelic depths for the alt alleles from vcf file.
  1. AF: AF or VAF from blood sample vcf file.
  1. DP: DP from vcf file.
  1. SAF: SAF info from vcf file.
  1. SAR: SAR info from vcf file.
  1. SampleID: sample ID or variant.
  1. Func.refGene: function annotation from refGene.
  1. ExonicFunc.refGene: exonic function annotation from refGene. (nonsynonymous SNV and synoymous SNV values need to be named as "nonsynonymous SNV" and"synonymous SNV")
  1. cosmic70: if the variant is exist in cosmic database. (empty value needs to be ".")
  1. tumor_AF: optional, AF or VAF from tumor sample vcf file.
  1. non_cancer_AF_popmax: optional, non cancer AF value from gnomad database.    
  1. Alt_dpGAP_PopFreq: optional, ALT population frequency from dpGAP databse.

To note: all empty value should be noted as "." in the merged text file.

```{r example_input_format}
# demo example of input merged file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# name of each variable
colnames(in_f)

# examples of each variable
head(in_f)
```


The function _vcf2input_ can convert raw or annotated VCF files into an input 
merged file for the main functions of qcCHIP. For raw VCF files, as they don't contain
the required information of _Func.refGene_, _ExonicFunc.refGene_, and _cosmic70_, 
converting with raw VCF files require users to manually add the corresponding columns 
to the merged file.

### Converting with raw VCF files

Below are the examples of how to convert raw VCF files into a merged text file.
For demo purpose, VCF files generated from 10 cancer patients are documented in
the package and used here.

```{r raw_vcf}
# Path to raw vcf files
raw_vcf_path<- system.file("extdata/raw_vcf",package="qcCHIP")

# Create data.frame of samples
sample_df <- data.frame(blood=list.files(raw_vcf_path)[! list.files(raw_vcf_path) %in% 
                                                         grep("*_tumor\\.vcf\\.gz",list.files(raw_vcf_path),value=T)],
                        tumor=grep("*_tumor\\.vcf\\.gz",list.files(raw_vcf_path),value=T))
head(sample_df)

# 1. Converting without paired tumor sample 
input_1 <- vcf2input(vcf_path = raw_vcf_path,
                   sample_list=sample_df$blood)
head(input_1)

# 2. Converting with paired tumor sample
input_2 <- vcf2input(vcf_path = raw_vcf_path,
                  sample_list=sample_df$blood,
                  tumor=T,
                  tumor_path=raw_vcf_path,
                  tumor_list=sample_df$tumor)
head(input_2)
```



### Converting with annotated VCF files

Currently, qcCHIP can take the following annotation information. These annotation can be generated by ANNOVAR tool.

  1. Function name, gene name, exonic function, AAchange, and GeneDetail information from refGene database;
  1. COSMIC info from COSMIC database (e.g. COSMIC70);
  1. Non cancer AF population max from gnomAD database.

Currently, function name, exonic function, and COSMIC are required annotation information by qcCHIP.

```{r annot_vcf}
# Path to annotated vcf files
annot_vcf_path<- system.file("extdata/annot_vcf",package="qcCHIP")

# Create data.frame of samples
sample_df <- data.frame(blood=list.files(annot_vcf_path)[! list.files(annot_vcf_path) %in% 
                                                      grep("*_tumor\\.hg38_multianno\\.vcf",list.files(annot_vcf_path),value=T)],
                        tumor=grep("*_tumor\\.hg38_multianno\\.vcf",list.files(annot_vcf_path),value=T))
head(sample_df)

# example of annotated VCF files
library(vcfR)
exp_vcf <- read.vcfR(system.file("extdata/annot_vcf","sample_1.hg38_multianno.vcf",package="qcCHIP"))

# check name of annotation information
strsplit(head(exp_vcf@fix,n=1)[,8],";")

# Converting annotated VCFs with paired tumor samples. Annotated with refGene, cosmic, and gnomAD.
input_3 <- vcf2input(vcf_path = annot_vcf_path,
                  sample_list=sample_df$blood,
                  tumor=T,
                  tumor_path=annot_vcf_path,
                  tumor_list=sample_df$tumor,
                  refGene=T,
                  refGene_func_name="Func.refGeneWithVer",
                  refGene_gene_name="Gene.refGeneWithVer",
                  refGene_Exonicfunc_name="ExonicFunc.refGeneWithVer",
                  refGene_AAchange_name=NA,
                  refGene_GeneDetail_name=NA,
                  cosmic=T,
                  cosmic_name="cosmic70",
                  gnomad=T,
                  gnomad_name="non_cancer_AF_popmax")

head(input_3)
```

The output file can be directly adopted by the main filtering and permutation functions.  

```{r annot_vcf_run}
# find CH with default metrics
ch_candidate <- CHIPfilter(input_3,tumor_sample=T)
head(ch_candidate)
```
 
# Basic Usage of The _qcCHIP_ Function

In this section, we use _qcCHIP_ function to perform permutation and evaluate the
effects of parameter cutoff values. Currently, _qcCHIP_ function supports optimization 
for five metrics: VAF, DP, SOR, SAF/SAR and cohort mutation prevalence. As demonstrated
in our manuscript, permutation consistency usually increases with higher parameter 
values with the exception that larger SOR decreases consistency. This is because that 
larger SOR means higher strand bias and lower reliability of mutation calls. 
In our manuscript, we demonstrated that the infection points of permutation consistency 
curves indicate the best parameter cut-offs. In practice, users should determine the 
infection points of permutation consistency curves: 1) with the 
increase of consistency stratified by VAF, DP, SAF/SAR and cohort mutation prevalence; 
2) with the decrease of consistency stratified by SOR. In addition, users are suggested to 
take full consideration across consistency, precision and recall curves to determine 
the best parameter cutoffs.  
 
Below we demo the permutation analysis and generate figures and summary for three parameters: 
VAF, DP, and cohort mutation prevalence. The figures and summary are used for determination of 
optimal parameter cutoffs used in the next section. 

### Run _qcCHIP_ with different VAF cutoffs 

The hypothesis is that a CH should hold a reliable VAF to be considered as true 
somatic mutation. This section demonstrates how different settings of minimum VAF 
affect the permutation consistency. Please refer to our manuscript for more detail.
For computing efficiency, we only demo with 10 samples here, which is less-powered 
compared to the results presented in our manuscript.

```{r qcCHIP_VAF}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/vaf_test")
vaf_permut <- qcCHIP(in_f,out_path = out_dir
                     ,metric_min = 0,
                     metric_step = 0.02,
                     metric_max = 0.1,
                     core=1,
                     show_info = F)

# example of comparision summary output
head(vaf_permut$summary_df)

# permutation consistency plot
vaf_permut$figs

```

### Run _qcCHIP_ with different DP cutoffs

The hypothesis is that a CH should hold a reliable DP to be considered as true 
somatic mutation. This section demonstrates how different settings of minimum DP 
affect the permutation consistency. Please refer to our manuscript for more detail.


```{r qcCHIP_DP}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/DP_test")
DP_permut <- qcCHIP(in_f,out_path = out_dir,permut_metrics = "DP", 
                    metric_min = 0,
                    metric_step = 10,
                    metric_max = 50,
                    core=1,
                    show_info = F)

# example of comparision summary output
head(DP_permut$summary_df)

# permutation consistency plot
DP_permut$figs

```

### Run _qcCHIP_ with different mutation prevalence cutoffs

The hypothesis is that a true CH shouldn't be over prevalent in a given cohort 
as CHs are relatively rare events. This section demonstrates how different settings 
of maximum mutation prevalence affect the permutation consistency. Please refer 
to our manuscript for more detail.


```{r qcCHIP_pop}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/population_test")
pop_permut <- qcCHIP(in_f,out_path = out_dir,permut_metrics = "population", 
                    metric_min = 0.05,
                    metric_step = 0.05,
                    metric_max = 0.5,
                    core=1,
                    show_info = F)

# example of comparison summary output
head(pop_permut$summary_df)

# permutation consistency plot
pop_permut$figs

```


# Basic Usage of The _CHIPfilter_ Function

Finally, we use _CHIPfilter_ to filter CH candidates based on a variety of quality 
metrics (detailed in the man page of _CHIPfilter_). The output will be a subset 
of input file which pass the QC filtering. It is recommended that users first run 
_qcCHIP_ function to determine optimal metric values. Nevertheless, users can  
pick their customized metric values and directly apply this function for CH filtering. 
Some features of _CHIPfilter_ are described below.

```{r CHIPfilter}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# exclude blacklist region 
bf_path<- system.file("extdata","demo_blacklist.bed",package="qcCHIP")

bl_f <- read.table(bf_path,sep = "\t",header=F)

# run default setting
out_1 <- CHIPfilter(in_f)

# use different metric values
out_2 <- CHIPfilter(in_f,max_percent=0.02,DP_min = 40,VAF_min=0.002,info=F)

# fitler with paired tumor sample
out_3 <- CHIPfilter(in_f,tumor_sample = T,tumor_VAF_min = 0.02,info=F)

# filter with gnomAD or dpGAP reference file
out_4 <- CHIPfilter(in_f,gnomad = F,dpGAP = F,info=F)

# filter with blacklist region
out_5 <- CHIPfilter(in_f,blacklist_f = bl_f,info=F)

# check the number of CHIP 
length(unique(out_1$mut_sample))
length(unique(out_2$mut_sample))
length(unique(out_3$mut_sample))
length(unique(out_4$mut_sample))
length(unique(out_5$mut_sample))

```



