---
title: "_qcCHIP_ User's Guide"
author: |
  | Xiang Liu <xiang.liu@moffitt.org>
  | Mingxiang Teng <mingxiang.teng@moffitt.org>
  | Department of Biostatistics and Bioinformatics
  | Moffit Cancer Center, Tampa, FL, USA
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true
  html_document:
    toc: true
    df_print: paged
graphics: true
vignette: "%\\VignetteIndexEntry{qcCHIP user's guide} %\\VignettePackage{qcCHIP} %\\VignetteEncoding{UTF-8}
  %\\VignetteEngine{knitr::rmarkdown}\n"
---

```{r, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  error = FALSE,
  warning = FALSE, 
  message = FALSE,
  cache = FALSE,
  comment = "#>",
  dev="png"
)
```

# Introduction
Clonal hematopoiesis (CH) is a molecular biomarker associated with various adverse outcomes in healthy and disease individuals. Detecting CHs usually involves genomic sequencing of individual blood samples followed by rigorous bioinformatics data filtering. We report an R package qcCHIP, a bioinformatics pipeline to identify CH from sequencing data by implementing a series of quality control filters and a permutation-based parameter optimization.

# Install package

Install _qcCHIP_ package via _devtools_.
```{r install}
library(devtools)
devtools::install_github("https://github.com/tenglab/qcCHIP.git",force=T)
```

# Getting Started

Load the package in R.

```{r library}
library(qcCHIP)
library(GenomicRanges)
```




# Preparing Input files

qcCHIP requires an annotated text file with specific column names:

All empty value should be noted as ".".

  1. Chr: chromosome of variant. Exp: chr1, chr2,chrX.
  1. Start: start posation of variant.
  1. End: end posation of variant.
  1. Ref: reference allele.
  1. Alt: alternative allele.
  1. TLOD: TLOD or Qual Info from vcf file.
  1. SOR: SOR Info from vcf file.
  1. AD_alt: Allelic depths for the alt alleles from vcf file.
  1. AF: AF or VAF from blood sample vcf file.
  1. DP: DP from vcf file.
  1. SAF: SAF info from vcf file.
  1. SAR: SAR info from vcf file.
  1. SampleID: sample ID or variant.
  1. Func.refGene: function annotation from refGene.
  1. ExonicFunc.refGene: exonic function annotation from refGene. (nonsynonymous SNV and synoymous SNV values need to be named as "nonsynonymous SNV" and"synonymous SNV")
  1. cosmic70: if the variant is exist in cosmic database. (empty value needs to be ".")
  1. tumor_AF: optional, AF or VAF from tumor sample vcf file.
  1. non_cancer_AF_popmax: optional, non cancer AF value from gnomad database.    
  1. Alt_dpGAP_PopFreq: optional, ALT population frequency from dpGAP databse.

There are two ways to prepare the input file.

  1. Using in-house converting tool to convert raw VCF or annotated VCF files.
  1. Manually create the input file with the columns described above.

```{r example_input_format}
# example input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# name of each variables
colnames(in_f)

# value format of each variables
head(in_f)
```

## Using VCF files as input

Tool vcf2input can convert raw or annotated VCF files to the input file that can be used for qcCHIP.

### Raw VCF files

Below are the examples of how to convert raw VCF files to input file.

```{r raw_vcf}
# Path of vcf files
raw_vcf_path<- system.file("extdata/raw_vcf",package="qcCHIP")

# Create data.frame of samples
sample_df <- data.frame(blood=list.files(raw_vcf_path)[! list.files(raw_vcf_path) %in% 
                                                         grep("*_tumor\\.vcf\\.gz",list.files(raw_vcf_path),value=T)],
                        tumor=grep("*_tumor\\.vcf\\.gz",list.files(raw_vcf_path),value=T))
head(sample_df)

# 1. No paired tumor sample and no annotation of other database
input_1 <- vcf2input(vcf_path = raw_vcf_path,
                   sample_list=sample_df$blood)
head(input_1)

# 2. With paired tumor sample and no annotation of other database
input_2 <- vcf2input(vcf_path = raw_vcf_path,
                  sample_list=sample_df$blood,
                  tumor=T,
                  tumor_path=raw_vcf_path,
                  tumor_list=sample_df$tumor)
head(input_2)
```

Note that, raw vcf files don't have the required information of Func.refGene, ExonicFunc.refGene, and cosmic70 for qcCHIP. Users need to manually adding the corresponding columns to the resulting table converted from raw VCF.

### Annotated VCF files

Currently, qcCHIP can take the following annotation information.

  1. Function name, gene name, exonic function, AAchange, and GeneDetail information from refGene database;
  1. COSMIC info from COSMIC database (e.g. COSMIC70);
  1. Non cancer AF population max from gnomad database.

Function name, exonic function, and COSMIC are required annotation information for qcCHIP.

```{r annot_vcf}
# Path of vcf files
annot_vcf_path<- system.file("extdata/annot_vcf",package="qcCHIP")

# Create data.frame of samples
sample_df <- data.frame(blood=list.files(annot_vcf_path)[! list.files(annot_vcf_path) %in% 
                                                      grep("*_tumor\\.hg38_multianno\\.vcf",list.files(annot_vcf_path),value=T)],
                        tumor=grep("*_tumor\\.hg38_multianno\\.vcf",list.files(annot_vcf_path),value=T))
head(sample_df)

# example of annotated VCF files
library(vcfR)
exp_vcf <- read.vcfR(system.file("extdata/annot_vcf","sample_1.hg38_multianno.vcf",package="qcCHIP"))

# check name of annotation information
strsplit(head(exp_vcf@fix,n=1)[,8],";")

# Annotated VCF of sample and paired tumor sample. Annotated with refGene, cosmic, and gnomad.
input_3 <- vcf2input(vcf_path = annot_vcf_path,
                  sample_list=sample_df$blood,
                  tumor=T,
                  tumor_path=annot_vcf_path,
                  tumor_list=sample_df$tumor,
                  refGene=T,
                  refGene_func_name="Func.refGeneWithVer",
                  refGene_gene_name="Gene.refGeneWithVer",
                  refGene_Exonicfunc_name="ExonicFunc.refGeneWithVer",
                  refGene_AAchange_name=NA,
                  refGene_GeneDetail_name=NA,
                  cosmic=T,
                  cosmic_name="cosmic70",
                  gnomad=T,
                  gnomad_name="non_cancer_AF_popmax")

head(input_3)
```

This output table is the correct format that can be directly used to find CH.  

```{r annot_vcf_run}
# find CH with default metrics
ch_candidate <- CHIPfilter(input_3,tumor_sample=T)
head(ch_candidate)
```
 
# Basic Usage of _qcCHIP_

In this section, we use _qcCHIP_ to test the results of select CHIP candidate with different setting of VAF, DP, or population. The resulting figures and comparision summary file will help user diceide the optimal VAF, DP, or population metric for their dataset. 

### *Run qcCHIP with change of minimum VAF*

This section demonstrates the usage of _qcCHIP_ when use different setting of minimum VAF. 

```{r qcCHIP_VAF}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/vaf_test")
vaf_permut <- qcCHIP(in_f,out_path = out_dir
                     ,metric_min = 0,
                     metric_step = 0.02,
                     metric_max = 0.1,
                     core=1,
                     show_info = F)

# example of comparision summary output
head(vaf_permut$summary_df)

# permutation consistency plot
vaf_permut$figs

```

### *Run qcCHIP with change of minimum DP*

This section demonstrates the usage of _qcCHIP_ when use different setting of minimum DP. 


```{r qcCHIP_DP}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/DP_test")
DP_permut <- qcCHIP(in_f,out_path = out_dir,permut_metrics = "DP", 
                    metric_min = 0,
                    metric_step = 10,
                    metric_max = 50,
                    core=1,
                    show_info = F)

# example of comparision summary output
head(DP_permut$summary_df)

# permutation consistency plot
DP_permut$figs

```

### *Run qcCHIP with change of maximum population percentage*

This section demonstrates the usage of _qcCHIP_ when use different setting of maximum population percentage. 

```{r qcCHIP_pop}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# create test directory
out_dir <- paste0(getwd(),"/population_test")
pop_permut <- qcCHIP(in_f,out_path = out_dir,permut_metrics = "population", 
                    metric_min = 0.05,
                    metric_step = 0.05,
                    metric_max = 0.5,
                    core=1,
                    show_info = F)

# example of comparision summary output
head(pop_permut$summary_df)

# permutation consistency plot
pop_permut$figs

```


# Basic Usage of _CHIPfilter_

In this section, we use _CHIPfilter_ to get the result of select CHIP candidate based on variety of selection matrics (detailed in the man page of _CHIPfilter_). The output will be a subset of input file which pass the selection. Users can directly use this function without runing _qcCHIP_. Some features of _CHIPfilter_ are described below.

```{r CHIPfilter}
# input file
input_path<- system.file("extdata","demo_input.txt",package="qcCHIP")
in_f <- read.table(input_path,sep="\t",header=T)

# blacklist region to exclude
bf_path<- system.file("extdata","demo_blacklist.bed",package="qcCHIP")

bl_f <- read.table(bf_path,sep = "\t",header=F)

# run default setting
out_1 <- CHIPfilter(in_f)

# change different metrics
out_2 <- CHIPfilter(in_f,max_percent=0.02,DP_min = 40,VAF_min=0.002,info=F)

# with paired tumor sample
out_3 <- CHIPfilter(in_f,tumor_sample = T,tumor_VAF_min = 0.02,info=F)

# with gnomad or dpGAP reference file
out_4 <- CHIPfilter(in_f,gnomad = F,dpGAP = F,info=F)

# with blacklist region
out_5 <- CHIPfilter(in_f,blacklist_f = bl_f,info=F)

# check the number of CHIP 
length(unique(out_1$mut_sample))
length(unique(out_2$mut_sample))
length(unique(out_3$mut_sample))
length(unique(out_4$mut_sample))
length(unique(out_5$mut_sample))

```



